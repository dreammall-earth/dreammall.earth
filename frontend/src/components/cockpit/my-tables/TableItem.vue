<template>
  <li class="table">
    <div
      class="table-info bg-dropdown-background animate-fade-out"
      :class="{ hidden: isShowingOptions }"
    >
      <span class="name">{{ props.name }}</span>
      <span class="subtitle">
        {{ $t('cockpit.myTables.memberCount', { count: props.memberCount }) }}
      </span>
    </div>
    <button
      class="action animate-fade-out"
      :class="{ hidden: isShowingOptions }"
      @click="openTable"
    >
      <v-icon class="camera-icon" icon="$camera" />
    </button>
    <button class="options" @click="toggleOptions">
      <v-icon
        class="options-icon"
        :class="{ turned: isShowingOptions }"
        icon="mdi mdi-dots-horizontal"
      />
    </button>
    <OptionsList :is-visible="isShowingOptions">
      <OptionButton class="trash" @click="deleteTable">
        <v-icon icon="mdi mdi-trash-can-outline" />
        {{ $t('cockpit.myTables.delete') }}
      </OptionButton>
      <OptionButton @click="invite">
        <v-icon icon="mdi mdi-account-plus-outline" />
        {{ $t('cockpit.myTables.invite') }}
      </OptionButton>
      <OptionButton @click="shareTable">
        <v-icon icon="mdi mdi-share-variant-outline" />
        {{ $t('cockpit.myTables.share') }}
      </OptionButton>
      <OptionButton @click="editTable">
        <v-icon icon="mdi mdi-pencil-outline" />
        {{ $t('cockpit.myTables.edit') }}
      </OptionButton>
    </OptionsList>
  </li>
</template>

<script lang="ts" setup>
import { navigate } from 'vike/client/router'
import { ref } from 'vue'

import OptionButton from '#components/cockpit/options-list/OptionButton.vue'
import OptionsList from '#components/cockpit/options-list/OptionsList.vue'

const props = defineProps<{
  id: number
  name: string
  memberCount: number
}>()

const isShowingOptions = ref(false)

const openTable = () => {
  navigate(`/table/${props.id}`)
}

const toggleOptions = () => {
  isShowingOptions.value = !isShowingOptions.value
}

const editTable = () => {
  // TODO implement
}

const invite = () => {
  // TODO implement
}

const shareTable = () => {
  // TODO implement
}

const deleteTable = () => {
  // TODO implement
}
</script>

<style scoped lang="scss">
.table {
  position: relative;
  display: flex;
  gap: 8px;
  align-items: center;
  justify-content: center;
}

.table-info {
  display: flex;
  flex-flow: column;
  justify-content: space-between;
  width: 100%;
  height: 42px;
  padding: 5px 24px;
  border-radius: 16px 0 0 16px;
}

.name {
  height: 18px;
  font-size: 14px;
  font-weight: bold;
}

.subtitle {
  font-size: 11px;
}

.action {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 41px;
  min-width: 41px;
  height: 42px;
  color: #fff;
  background-color: #f09630;
  border-radius: 0 16px 16px 0;
}

.camera-icon {
  transform: scale(0.8);
}

.options-icon {
  transition: transform 0.3s;
  transform: rotate(0);

  &.turned {
    transform: rotate(90deg);
  }
}

.trash {
  background-color: #d02f44;
}

.animate-fade-out {
  opacity: 1;
  transition: opacity 0.3s;
}

.hidden {
  opacity: 0;
}
</style>
