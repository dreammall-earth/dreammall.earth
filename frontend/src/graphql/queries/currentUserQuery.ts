import { gql } from 'graphql-tag'

export const currentUserQuery = gql`
  query {
    currentUser {
      id
      name
      username
      room {
        id
        name
        public
        users {
          id
          role
          name
          username
        }
      }
    }
  }
`
