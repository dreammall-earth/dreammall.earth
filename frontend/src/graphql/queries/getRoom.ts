import { gql } from 'graphql-tag'

export const getRoom = gql`
  query {
    getRoom
  }
`
