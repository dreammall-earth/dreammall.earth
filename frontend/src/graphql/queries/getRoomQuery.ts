import { gql } from 'graphql-tag'

export const getRoomQuery = gql`
  query {
    getRoom
  }
`
