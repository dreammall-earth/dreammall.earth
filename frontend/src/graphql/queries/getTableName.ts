import { gql } from 'graphql-tag'

export const getTableNameQuery = gql`
  query getTableName($tableId: String!) {
    getTableName(tableId: $tableId)
  }
`
