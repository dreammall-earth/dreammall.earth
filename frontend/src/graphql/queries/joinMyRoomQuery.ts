import { gql } from 'graphql-tag'

export const joinMyRoomQuery = gql`
  query {
    joinMyRoom
  }
`

export type JoinMyRoomQueryResult = {
  joinMyRoom: string
}
