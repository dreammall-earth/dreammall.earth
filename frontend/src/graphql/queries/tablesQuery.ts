import { gql } from 'graphql-tag'

export const tablesQuery = gql(`{
  tables {
    id
    name
    public
    users {
      id
      name
      role
      username
    }
  }
}`)
