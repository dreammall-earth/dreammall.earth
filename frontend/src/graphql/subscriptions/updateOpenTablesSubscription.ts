import { gql } from 'graphql-tag'

export const updateOpenTablesSubscription = gql`
  subscription ($username: String!) {
    updateOpenTables(username: $username) {
      meetingID
      meetingName
      startTime
      participantCount
      attendees {
        fullName
      }
      joinLink
    }
  }
`
