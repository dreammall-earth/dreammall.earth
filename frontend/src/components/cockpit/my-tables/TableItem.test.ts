import { provideApolloClient } from '@vue/apollo-composable'
import { mount } from '@vue/test-utils'
import { createMockClient, createMockSubscription, IMockSubscription } from 'mock-apollo-client'
import { navigate } from 'vike/client/router'
import { describe, expect, it, vi } from 'vitest'

import { currentUserQuery } from '#src/graphql/queries/currentUserQuery'
import { tablesQuery } from '#src/graphql/queries/tablesQuery'
import { updateOpenTablesSubscription } from '#src/graphql/subscriptions/updateOpenTablesSubscription'

import TableItem from './TableItem.vue'

vi.mock('vike/client/router')

const currentUserQueryMock = vi.fn()
const updateOpenTablesSubscriptionMock: IMockSubscription = createMockSubscription()
const tablesQueryMock = vi.fn()

const mockClient = createMockClient()

mockClient.setRequestHandler(
  currentUserQuery,
  currentUserQueryMock.mockResolvedValue({
    data: {
      currentUser: {
        id: 666,
        name: 'Current User',
        username: 'currentUser',
        table: null,
      },
    },
  }),
)

mockClient.setRequestHandler(updateOpenTablesSubscription, () => updateOpenTablesSubscriptionMock)

mockClient.setRequestHandler(
  tablesQuery,
  tablesQueryMock.mockResolvedValue({
    data: {
      tables: [
        {
          id: 1,
          name: 'My Table',
          public: false,
          users: [
            {
              name: 'Current User',
              username: 'currentUser',
              role: 'MODERATOR',
            },
            {
              name: 'Bibi Bloxberg',
              username: 'bibi',
              role: 'VIEWER',
            },
            {
              name: 'Peter Lustig',
              username: 'peter',
              role: 'VIEWER',
            },
          ],
        },
      ],
    },
  }),
)

provideApolloClient(mockClient)

describe('Table Item', () => {
  const Wrapper = (
    props = {
      id: 1,
      name: 'Table Name',
      memberCount: 4,
    },
  ) =>
    mount(TableItem, {
      props,
    })

  it('renders', () => {
    const wrapper = Wrapper()
    expect(wrapper.element).toMatchSnapshot()
  })

  describe('camera icon', () => {
    it('navigates to table page when clicked', async () => {
      const wrapper = Wrapper()
      await wrapper.find('.camera-icon').trigger('click')
      expect(navigate).toHaveBeenCalledWith('/table/1')
    })
  })

  describe('options button', () => {
    it('opens options when clicked', async () => {
      const wrapper = Wrapper()
      await wrapper.find('button.options').trigger('click')
    })
  })
})
