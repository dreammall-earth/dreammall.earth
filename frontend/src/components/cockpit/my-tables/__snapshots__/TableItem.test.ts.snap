// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table Item > renders 1`] = `
<li
  class="table"
  data-v-dca20f6c=""
>
  <div
    class="table-info bg-dropdown-background animate-fade-out"
    data-v-dca20f6c=""
  >
    <span
      class="name"
      data-v-dca20f6c=""
    >
      Table Name
    </span>
    <span
      class="subtitle"
      data-v-dca20f6c=""
    >
      $t('cockpit.myTables.memberCount')
    </span>
  </div>
  <button
    class="action animate-fade-out"
    data-v-dca20f6c=""
  >
    <i
      aria-hidden="true"
      class="v-icon notranslate v-theme--light v-icon--size-default camera-icon"
      data-v-dca20f6c=""
    >
      <svg
        fill="none"
        height="17"
        viewBox="0 0 21 17"
        width="21"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.68844 12.4824H9.66834V9.51172H12.6382V7.53125H9.66834V4.56055H7.68844V7.53125H4.71859V9.51172H7.68844V12.4824ZM2.73869 16.4434C2.19422 16.4434 1.72812 16.2494 1.34038 15.8616C0.952654 15.4738 0.758789 15.0075 0.758789 14.4629V2.58008C0.758789 2.03545 0.952654 1.56921 1.34038 1.18137C1.72812 0.79353 2.19422 0.599609 2.73869 0.599609H14.6181C15.1626 0.599609 15.6287 0.79353 16.0164 1.18137C16.4041 1.56921 16.598 2.03545 16.598 2.58008V7.03613L20.5578 3.0752V13.9678L16.598 10.0068V14.4629C16.598 15.0075 16.4041 15.4738 16.0164 15.8616C15.6287 16.2494 15.1626 16.4434 14.6181 16.4434H2.73869ZM2.73869 14.4629H14.6181V2.58008H2.73869V14.4629Z"
          fill="currentColor"
        />
      </svg>
    </i>
  </button>
  <button
    class="options"
    data-v-dca20f6c=""
  >
    <i
      aria-hidden="true"
      class="mdi mdi-dots-horizontal mdi v-icon notranslate v-theme--light v-icon--size-default options-icon"
      data-v-dca20f6c=""
    />
  </button>
  <transition-group-stub
    appear="false"
    css="true"
    data-v-01c78171=""
    data-v-dca20f6c=""
    name="options"
    persisted="false"
  >
    <!--v-if-->
  </transition-group-stub>
</li>
`;
